File: language/benchmarks/tokens/algorithms/levenshtein-distance/levenshtein-distance.ts
function levenshtein(left: string, right: string): number {
  return levenshteinFrom(left, 0, right, 0);
}

function levenshteinFrom(left: string, leftIdx: number, right: string, rightIdx: number): number {
  if (leftIdx === left.length) return right.length - rightIdx;
  if (rightIdx === right.length) return left.length - leftIdx;
  if (left.charAt(leftIdx) === right.charAt(rightIdx)) {
    return levenshteinFrom(left, leftIdx + 1, right, rightIdx + 1);
  }
  return 1 + min3(
    levenshteinFrom(left, leftIdx + 1, right, rightIdx),
    levenshteinFrom(left, leftIdx, right, rightIdx + 1),
    levenshteinFrom(left, leftIdx + 1, right, rightIdx + 1)
  );
}

function min2(a: number, b: number): number {
  if (a <= b) return a;
  return b;
}

function min3(a: number, b: number, third: number): number {
  return min2(min2(a, b), third);
}

File: language/benchmarks/tokens/algorithms/levenshtein-distance/levenshtein-distance.py
def levenshtein(left: str, right: str) -> int:
    return levenshtein_from(left, 0, right, 0)


def levenshtein_from(left: str, left_idx: int, right: str, right_idx: int) -> int:
    if left_idx == len(left):
        return len(right) - right_idx
    if right_idx == len(right):
        return len(left) - left_idx
    if left[left_idx] == right[right_idx]:
        return levenshtein_from(left, left_idx + 1, right, right_idx + 1)
    return 1 + min3(
        levenshtein_from(left, left_idx + 1, right, right_idx),
        levenshtein_from(left, left_idx, right, right_idx + 1),
        levenshtein_from(left, left_idx + 1, right, right_idx + 1),
    )


def min2(a: int, b: int) -> int:
    if a <= b:
        return a
    return b


def min3(a: int, b: int, third: int) -> int:
    return min2(min2(a, b), third)
